### PR TITLE
MGMT-132: Supporting migrations

### PR DESCRIPTION
**Pull request by dbhynds**
## Summary of Changes

**Background**

- This conversation began in this pull request: https://github.com/packbackbooks/lti-1-3-php-library/pull/98
- IMSGlobal's description of how this should function can be found here: https://www.imsglobal.org/spec/lti/v1p3/migr

**Summary of the problem**

While validating an LTI Message Lauch, the library throws an exception if it can't find a deployment. However, when migrating, if the launch contains a `oauth_consumer_key_sign`, the tool may try to upgrade the launch from 1.1 to 1.3. In that case, an deployment should 

We're essentially muddying the waters in terms of separation of concerns: validation and migration. We need to validate the LTI launch message before we migrate, but we also need a legitimate path for migration, while still also ultimately validating the deployment provided by the launch message.

**How this would be used**

A tool provider would implement their own migration code in the database service. Then in the launch you would simply do:

```php
$message->initialize($request->all());
```

## Testing

- [x] I have added automated tests for my changes
- [x] I ran `composer test` before opening this PR
- [x] I ran `composer lint-fix` before opening this PR

### CODE REVIEW

**Comment by dbhynds** [comment]
@erikdonohoo Tagging you for your feedback.

File: src/LtiMessageLaunch.php
Comment (erikdonohoo) [inline src/LtiMessageLaunch.php:162]: Where is `hasMigrationStrategy()` defined?

File: src/LtiMessageLaunch.php
Comment (erikdonohoo) [inline src/LtiMessageLaunch.php:165]: In order to verify this, you need to be passed both `oauth_consumer_key_sign` and `oauth_consumer_key`. These properties will be under the 1.1 migration claim, and not on the top level of the launch data.

Its also possible to not send these values, so you can't assume they are there (referring to the key on the array if its not there will blow up).

**Comment by erikdonohoo** [comment]
Left a few notes on this. Anyway I could take a stab at modifying some of this?

**Comment by dbhynds** [comment]
Yep! I'll look into getting you access tomorrow.

File: src/LtiMessageLaunch.php
Comment (dbhynds) [reply src/LtiMessageLaunch.php:165]: I threw this together from memory without looking to see where these actually come from or putting in any safeguards. (This definitely isn't meant to be production-ready). Both of your suggestions here sound right to me.

File: src/LtiMessageLaunch.php
Comment (dbhynds) [reply src/LtiMessageLaunch.php:162]: It's not and I wasn't sure where it should be. It depends on how we end up defining the migration process.

**Comment by erikdonohoo** [comment]
Hey there @dbhynds . I think I found a nice way to handle this while fully preserving backwards compatibility. Let me know what you think of these changes.

**Comment by erikdonohoo** [comment]
I see @Watercycle is an assigned reviewer of this. Pinging as well.

File: .vscode/launch.json
Comment (dbhynds) [inline .vscode/launch.json:19]: Let's add this to the `.gitignore`

**Comment by dbhynds** [comment]
@erikdonohoo have you tested this out IRL yet? At the very least, it looks like we're mixing types for `$this->db` on the `LtiMessageLauch`, which I suspect will cause issues.

Also, could you give me a rough idea of what your implementation of the `IMigrationDatabase` looks like? I don't need to see anything proprietary, but just enough to give me a sense for the main things it's doing.

File: src/LtiMessageLaunch.php
Comment (dbhynds) [inline src/LtiMessageLaunch.php:156]: Does IMSGlobal provide guidance on how failures should be handled? As this is currently implemented, a mismatch would throw an exception which would be returned to the user. This means all LTI launches would fail until it was corrected. That seems bad.

**Comment by erikdonohoo** [comment]
> @erikdonohoo have you tested this out IRL yet? At the very least, it looks like we're mixing types for `$this->db` on the `LtiMessageLauch`, which I suspect will cause issues.
> 
> Also, could you give me a rough idea of what your implementation of the `IMigrationDatabase` looks like? I don't need to see anything proprietary, but just enough to give me a sense for the main things it's doing.

Sure thing. Yes this does work in real life, and the unit tests prove that. We aren't mixing types, as they are both interfaces. You can check to see if the actual real life DB class implements either class. We are only checking for migrations if the new interface was actually implemented by the db class. This makes it completely backwards compatible. People who do not implement the new interface will see no change in behavior.

As far as how those methods would be implemented, it depends on how each tool stored data about 1.1 installations in their system. But the idea is for `getMatchingLti1p1Install` you look at all the data coming in from the launch and see if you have a matching 1.1 install based on the 1.1 claim or other claims in the launch. It will all be up to the tool to figure that out. In our case, we used things like the tool_consumer_instance_guid, combined with some other properties to uniqely identify the "organizations" representing a 1.1 launch.

For `migrateFromLti1p1`, the goal is to add a LtiDeployment into your system automatically. BY the time this method is called, a passed oauth consumer key sign has already been automatically verified, so it is safe to "auto provision" the 1.3 deployment. This code will create that deployment, as well as do any other rebinding or associating needed to the 1.1 installation. Again, this is all going to be dependent on how an individual tool stored that data in their own system.

@dbhynds Let me know what other questions you have on this

**Comment by dbhynds** [comment]
Ah, I misunderstood your approach with the interfaces, but I get it now. I like what you did.

File: src/LtiMessageLaunch.php
Comment (dbhynds) [reply src/LtiMessageLaunch.php:156]: Not trying to be pedantic, but I keep going back to the failure states here. I want to make sure we never fail an otherwise valid launch just because something related to the migration is incorrect. Consider:

A 1.3 integration has been successfully set up with several valid, existing deployments. A school admin sets things up to do a migration from 1.1 and messes something up with the oauth key/secret. If a user from an existing deployment launches, they shouldn't see an error message. We'd only want to serve an error for launches to deployments we don't recognize.

Comment (erikdonohoo) [reply src/LtiMessageLaunch.php:156]: If you already have a working 1.3 integration, then migration is no longer an option. Otherwise, the tool would need to support rebinding all the data and things created in the 1.3 migration to work again in the 1.1 integration. The point is if you are a tool supporting this, you block them before they start using it. Not that you block them when they are already using it. Even if they did do something wrong with the keys/secrets however, this STILL wouldn't get triggered, as it only gets triggered if the deployment_id in the launch isn't already found in the system. Migration can only occur if that deployment is not already registered in the tool. Does that make sense @dbhynds ? The `validateDeployment()` method only tries to do this code that can throw this error if it didn't find the deployment in the first attempt.

File: src/Lti1p1Installation.php
Comment (dbhynds) [inline src/Lti1p1Installation.php:25]: @erikdonohoo One final question (hopefully). In what circumstance would you have multiple potential keys to try to match against? I would assume that some combination of the `oauth_consumer_key`, `iss` and/or `aud` would be enough to definitively narrow it down to a single key when querying the database. Is that not the case?

Ideally I'd like to remove the `for` loop here, but I don't have enough context into how the 1.1 keys get queried.

Comment (erikdonohoo) [reply src/Lti1p1Installation.php:25]: The short answer is key rotation. If you want to support rotating keys, you need to support multiple keys for a 1.1 installation. Having that be an array, enables that. Does that make sense @dbhynds ?

Comment (dbhynds) [reply src/Lti1p1Installation.php:25]: Boo. Yeah, that makes sense. Alas, the `for` loop shall remain!

File: src/LtiMessageLaunch.php
Comment (dbhynds) [inline src/LtiMessageLaunch.php:190]: @erikdonohoo Does this match your expected implementation?

**Comment by dbhynds** [comment]
@Watercycle this is ready for you when you get back

File: src/LtiMessageLaunch.php
Comment (erikdonohoo) [inline src/LtiMessageLaunch.php:178]: Let me describe a situation here.  A user used 1.1 and wants to migrate. Their LMS however does not support the 1p1 migrate claim or the oauth_consumer_key_sign. Despite this however, using things like the `toolPlatform->guid` claim, we can reasonably infer that this user NEEDS to migrate, even though they didn't pass the `oauth_consumer_key`. In that case, I, the tool provider, need a different error out of packback. They "can't" migrate, but they "should" have migrated. This helps me in the messaging I will show them, as well as handling manually creating the deployment, to ensure the re-binding can occur manually on my side. This migrate function should change slightly.

1. The `canMigrate` function could be renamed to `implementsMigrationDb()` and check if the tool provider implemented code for migration, and if findLti1p1Keys returns any keys. If the deployment is set, or they didn't implement migration, or there is no key to migrate, then just return, as migration either isn't necessary, or isn't supported.
2. If we found a key, but the system didn't pass the key_sign or the consumer_key, then we shuld throw an error saying the KEY_SIGN or KEY is missing.
3. If we have the keys, but the signatures don't match, we should throw a KEY_SIGN_MISMATCH error.
4. Otherwise, we should do the migration.

**Comment by dbhynds** [comment]
Here are these changes in our CI pipeline: https://gitlab.com/packback/questions/-/pipelines/1039934019

File: src/LtiMessageLaunch.php
Comment (dbhynds) [reply src/LtiMessageLaunch.php:178]: Let me make sure I understand the scenario you described in the first paragraph:

1. A user performs an LTI 1.3 launch.
2. The LMS does not support migrations, and therefor the `oauth_consumer_key` is not included in the launch.
3. Despite this, there is a matching LTI 1.1 key that could be migrated (i.e., calling `findMatchingLti1p1Keys()` returns several results), but cannot be automatically due to the missing claims.

In that case, how would you know whether a user has already migrated or not? What in the system indicates "the user should migrate" vs "the user already migrated"?

File: src/LtiMessageLaunch.php
Comment (snake) [reply src/LtiMessageLaunch.php]: I've just stumbled across this one and would like to echo the initial sentiments around failure states and blocking launches and really ensure sure that migration is either an optional feature of this library or there is sufficient flexibility around what will result in a launch-blocking exception. 

We (Moodle) use this library in our tool implementation and because we support 1.1 and 1.3 coexistence, we don't have that same restriction that "_Migration can only occur if that deployment is not already registered in the tool_". It's possible to have a registration, complete with deployment id, set up and also process migration claims from the platform (which, for us, is a way to identify the same user in the tool, rather than a client, since we're not multi-tenant). We currently do outside of this library. We'd like to ensure that the library won't start throwing blocking errors during launch because of any number of missing claims or tool consumer data. 

We've already had reports of Canvas sending migration claims during launches (See https://tracker.moodle.org/browse/MDL-78945) and our migration code failing to find the relevant 1.1 consumer mapping and blocking the launch. We'd like to retain the option to not block launches entirely in such cases. Mainly, because whether or not erroneous migration claim data is present is outside of our control. In this case, the user wasn't even expecting a migration to occur and was entirely unable to use the tool at this point.

Comment (dbhynds) [reply src/LtiMessageLaunch.php]: @snake and @erikdonohoo ... how do y'all each handle the scenario where a user launches with a `deployment_id` that isn't recognized by your system. This package throws an exception stating that we couldn't find a deployment ID. That functionality existed when I took this repo over from IMS Global. However, looking through the docs, I'm not sure if that's the correct way to handle things. I've seen some comments that lead me to believe that it's safe to auto-create deployments if the tool doesn't find an existing one (as opposed to throwing an exception).

TLDR: should we change it so that--if a deployment ID is not found--the tool automatically creates the deployment ID instead of having the package throw an exception?

I ask because that exception and logic around validating deployments is what originally prompted the PR from Erik. If it's safe to auto-create deployments, I'm thinking it might be worth breaking off an entirely separate service for migrations--one that can exist independently from a launch. This would leave the launch process largely unchanged, but give tools flexibility and structure for how to implement migrations. (A tool could still implement automatic migrations as part of the launch process, but you wouldn't be tied to that.)

Comment (snake) [reply src/LtiMessageLaunch.php]: @dbhynds, for us (and noting we're quite different to multi-tenant saas tools out there), we require the deployment id to be pre-registered with us beforehand. So, in a normal launch, where we're not dealing with migration claims, failing when we see an unknown deployment id is desirable for us. Because we require the deployment id upfront we're also able to apply the same logic during launches which include the migration claim. I think if we could have more control over the 'should migrate' decision (aka the decision that's currently being made by the library in the proposed migrate() method, then we'd be able to use this. At present, that code will skip launches which include a known deployment and which contain a migration claim.

Sidenote: You'll see other libs which also validate deployment during the launch, in the tool code - e.g. https://github.com/oat-sa/lib-lti1p3-core/blob/e464bdd161a8d47a572af0867ec45ded2d54c21a/src/Message/Launch/Validator/Tool/ToolLaunchValidator.php#L250. Of course, there are probably ways to use said libraries and get around that, for specific cases like first-launch or migration.

I'm sure Erik will be able to expand on this next bit, but I believe a lot of other tool providers support a 'self service' kind of deployment-to-client mapping process, meaning they are able to take a launch and bind the as-yet-unseen deployment to one of their clients on the fly, during that first launch (take them through login, etc?). I'm not 100% sure on that process having not used it. After that, the deployment is known to the tool and is mapped. I believe that the plan for tools like this would be to handle migration in much the same way, except that instead of self service to find the client, we'd be able to look them up via the 1.1 details in the migration claim. Either way, the tool won't know the deployment id upfront during the first launch.

Because deployment represents a client in these cases, I don't know that we'd be able to automatically create the deployment as part of the launch process. Or rather, we'd perhaps want to 'do something' before we green light the creation of the deployment. In the migration case, we're wanting to check a few things like the consumer key and signature before doing this. In the non-migration case, I'm guessing the tool would want to make sure the user is signed in etc first, and make sure it can identify a client, before then making that decision to approve the deployment and map it to said client. I'm hoping Erik can add more on this.

Comment (erikdonohoo) [reply src/LtiMessageLaunch.php]: We support all major LMSs, and even homegrown LMSs. Basically anything that can do an LTI 1.1 launch, we work with. So we have a wide variety of different types of data in our system tied to the LTI spec. Our issue is ultimately this:

We have a tool that allows students to submit videos and get feedback from teachers on those videos. This comes with all kinds of other features and history that a user has access to. We need to ensure that all our existing customers who move to 1.3, have their 1.1 users bound to the new 1.3 users we make so they are in fact one and the same. So they maintain access to the content they have been creating for years with us using 1.1. If that re-binding does not take place at the beginning, then those users begin 1.3 with a new "user" in our system that is essentially invisible to them. Our tool doesn't require users to login to an account to begin with, as that account is strictly tied to the information coming in via the 1.1 launch. Being tied to a specific key/secret is where the security of the launches come in, not from them authorizing through some separate login on our side.

With that in mind, this is why on initial launch it is so important for us to handle that rebinding. The launch itself provisions entities on our side, and if we don't know what to do with the IDs, new things will be created in our DB that are not associated with the original 1.1 user that person had.

With 1.3, we intend to block all new launches with an unknown deployment as well. And I believe that is the correct way to do things according to the IMS global spec. We will have an internal tool that saves all the info that we could have used to auto-provision that deployment, and allow our support teams to verify the identify of the launcher and get that deployment_id in our system that way. After that, launches will work successfully. It is also at that moment where we will determine if they are an existing 1.1 customer or not and "migrate" them by mapping their existing 1.1 organization to the new params coming in for 1.3 launch. That way, they can launch 1.1 or 1.3 into the same content and as long as the ids they are passing are the same or mapped in the migration claim, it will always work.

As you know, almost no LMSs support the migration claim. IMS global documented the state of it here. https://www.imsglobal.org/lti-platform-migration-specifics They also produced a video with details on each LMSs support plans for migrations here https://www.youtube.com/watch?v=sg0dhf5j2xs

The hope was if the oauth_consumer_key_sign was passed, that this could be a short cut for customers who want to move to 1.3 if their LMS supports that param. It would allow us to "verify their identity" in the launch automatically and do the "re-binding" that IMS global speaks about in the video securely. None of this matters if your tool doesn't persist things between launches for your users and make things accessible between launches for them based on the user_id in the launch params (sub). But for us, we do that. And the only way to securely move them over with no intervention, would be this claim. Any migrations without this claim are technically a security risk vulnerable to social engineering. We are working with Blackboard and Canvas on this and have open security issues with them on the lack of having this for 1.1 migrations.

Hopefully this didn't sound like just straight on blabbering, but we've definitely been thinking about and working on this for a while and want to get it right for our customers when we turn on 1.3 support.

If we think it would be valuable, maybe we could schedule a group call between some of the interested parties here to make sure whatever is added into packback for migration supports everyones needs. We certainly don't want to cause issues for others that are ahead of us on 1.3 at this point.

Comment (erikdonohoo) [reply src/LtiMessageLaunch.php]: @dbhynds I just looked over your most recent changes, particularly to `shouldMigrate` by adding it to the DB. I like this as it allows the tool providers more control over when to do the migration. I am satisfied with this and think this would work fine for our needs. I would be interested in hearing what @snake has to say on this as well.

Comment (dbhynds) [reply src/LtiMessageLaunch.php]: I think hopping on a call might be helpful. Thanks for suggesting! I'm assuming everyone is located in the US, but correct me if I'm wrong. Here's a doodle for us to coordinate a time to talk. https://doodle.com/meeting/participate/id/bq2LGDGd

@snake @erikdonohoo

Comment (snake) [reply src/LtiMessageLaunch.php]: @dbhynds @erikdonohoo: that recent change also suits our (albeit probably obscure) use case. As long as we can make that decision about whether to continue with the migration, that ticks the box for us. 

Happy to jump on a call but I'm GMT+08, so it's usually an early morning 8am or 8pm affair for meetings with the states.

**Comment by dbhynds** [comment]
@snake @erikdonohoo I still have a few questions that will just be quicker to talk through live. Send me an email at davo@packback.co (easier than github comments) and let's find a time to chat. I'm thinking the best time to meet will be:
- `8 am` (snake)
- `6 pm` (erik)
- `7 pm` (me)

Next week I'm free Wed and Thurs evening (i.e. Thurs or Fri morning for snake)

File: src/LtiMessageLaunch.php
Comment (erikdonohoo) [reply src/LtiMessageLaunch.php]: @snake @dbhynds I've updated my github notification settings but for whatever reason I'm still not getting notified when we reply to each other either. I'm still available for that call if we feel its necessary, but I agree with @snake the most recent chagnes made will most likely satisfy everyone's needs. (At least as far as I can see)

File: src/LtiMessageLaunch.php
Comment (dbhynds) [inline src/LtiMessageLaunch.php:525]: @snake @erikdonohoo before I merge this, I just want to verify a few things:

1) Is this logic `canMigrate()` accurate here? Or should it be `shouldMigrate()`?
2) I've currently got `matchingLti1p1KeyExists()` marked private. Should it be public? I wasn't sure if you would need it as part of checking `$this->db->shouldMigrate()`

Comment (snake) [reply src/LtiMessageLaunch.php:525]: 1. For our case, shouldMigrate() is really most useful when we've already passed validation of the launch data (i.e. deployment is found and valid). At this point, shouldMigrate() allows check whether we've migrated already, and also would allow us to quickly skip migration again on subsequent launches (because, of course, migrate() is still being called on every launch). So, for us, what you've got works fine (as would shouldMigrate()) since we're dealing with the case where we DO have a known deployment id. I think it's Erik's scenario that's more prevalent though and which really hits this conditional. In that case, if you don't have a deployment mapped and are unable to migrate (because of an incompatible DB type), you'd want to fail here. 

I can't think of what calling shouldMigrate() would add. It seems it might support the following:
- I don't know this deployment id
- I am equipped to migrate (i.e. I have an IMigrationDatabase)
- but I want to invalidate the launch before attempting migration

I can't think of a situation like that, personally. Maybe others can?

2. I think we would probably use the database method findLti1p1Keys() if we needed to do that, so no problem leaving it as is. As long as we have the launch data, and an opportunity to control whether migration happens, we're happy.

**Review by Watercycle — APPROVED** [review]
Had a few inconsequential code & documentation comments. Otherwise, no concerns here.

As noted in a comment, please let me know if there's any explicit QA you'd like performed. We're not currently doing LTI 1.1 migrations like this in our application; but, based on what I'm seeing here, this is a flexible architecture to do so that provides reasonable scaffolding instead of having folks roll their own solutions.

P.S. Great discussions and awesome work on this PR!

File: src/ImsStorage/ImsCache.php
Comment (dbhynds) [reply src/ImsStorage/ImsCache.php:9]: I've got a separate PR open where I'm going to go through and do all of that.

**Comment by dbhynds** [comment]
@Watercycle I've made a few changes. I'm also going to go back and update documentation in the wiki after this merges.

**Pull request merged**
